### ISSUE TITLE
BPi Pro | Update to 4.9 kernel

### ISSUE THREAD

**Issue by Fourdee** [MichaIng/DietPi#732]

**Comment by Fourdee** [comment]
Mainline 4.9: http://dietpi.com/downloads/testing/4.9_image_tests/DietPi_BananaPiPro-armv7-(Jessie).7z

**Comment by Fourdee** [comment]
- 🈯️ onboard wifi
- 🈴 onboard sound (3.5mm/HDMI). Strange, https://linux-sunxi.org/Linux_mainlining_effort#Status_Matrix reports ```audio codec``` made it to kernel 4.4. Surely AC97 isn't required?

**Comment by gpo123** [comment]
Sound on 3,5mm jack is working on my BananaPi (not PRO) with self compiled Armbian kernel 4.9. Had to reset alsamixer settings though. Settings:  Left Mixer Left DAC : muted  / Power Amplifier DAC : unmuted / Power Amplifier Mixer :muted / Power Amplifier Mute : unmuted / Right Mixer Left DAC : muted / Right Mixer Right DAC : muted. HDMI ouput seems to be gone, but I don't care for my setup (remember what I said about reset /default values of soundcards ;-) )

I have a little problem with configuring the ondemand governor:  sample rate setting is not accessible due to /sys changes, CPU stays at max-frequency. Will investigate further.

**Comment by Fourdee** [comment]
@gpo123 

Thanks for info 👍 

> Sound on 3,5mm jack is working on my BananaPi (not PRO) with self compiled Armbian kernel 4.9. Had to reset alsamixer settings though

Yeah, i did play with the mixer settings, but had no joy on both hdmi and 3.5mm. I'll replicate your amixer settings and re-test.

**Comment by gpo123** [comment]
ok, got the ondemand bug: line 1195 in dietpi-config should read: MIN_VALUE=$(( $(cat /sys/devices/system/cpu/cpufreq/ondemand/**min_sampling_rate**) / 1000 )). After fiddling with the settings CPU is now scaling up and down nicely. According to [http://linux-sunxi.org/Cpufreq](url) you should also set **io_is_busy** to **1**. Had good results on my cubietruck.
Do you intend to support all governors available on 4.9 ?

**Comment by Fourdee** [comment]
@gpo123 

Are you Jórg by any chance 😃 ?

https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-cpuinfo#L153

We'll need to add an array of possible locations for:
```
/sys/devices/system/cpu/cpufreq/ondemand/min_sampling_rate
/sys/devices/system/cpu/cpu$i/cpufreq/scaling_min_freq
```

> Do you intend to support all governors available on 4.9 ?

All, excluding ones not available on mainline (eg:```interactive```)

> io_is_busy to 1

We'll need to add that in here: https://github.com/Fourdee/DietPi/blob/master/dietpi/dietpi-cpu_set#L99-L123

**Comment by gpo123** [comment]
Yes, I'm Jörg. I will try to make the approbiate changes to dietpi-config to support both kernels on BananaPi (PRO) later on.

**Comment by Fourdee** [comment]
@gpo123 

Yep, still no joy with BPi Pro. Tried various combinations, no sound:
![image](https://cloud.githubusercontent.com/assets/7077026/22372777/82f71698-e496-11e6-9fca-47de16642c1f.png)

**Comment by gpo123** [comment]
Hi Dan,

These are my settings with alsamixer -V all:

![image](https://cloud.githubusercontent.com/assets/9412171/22374576/010b5a16-e4a7-11e6-8ccc-4119b578e90c.png)


Should I upload my kernel / Armbian deb / kernel config somewhere ?

Regards Jörg


Am 27.01.2017 um 14:43 schrieb Dan:

**Comment by Fourdee** [comment]
@gpo123 

> alsamixer -V

```
alsamixer -V -c 0
```

Legend 👍. Now it works on 3.5mm!

![image](https://cloud.githubusercontent.com/assets/7077026/22377776/00417cd8-e4ab-11e6-9ece-3a8f37d0e134.png)

Ok so, 3.5mm, we need to set:
```
amixer sset 'Left Mixer Left DAC' off
amixer sset 'Power Amplifier' 62
amixer sset 'Power Amplifier DAC' on
amixer sset 'Power Amplifier Mixer' off
amixer sset 'Power Amplifier Mute' on
amixer sset 'Right Mixer Left DAC' off
amixer sset 'Right Mixer Right DAC' off
```

**Comment by gpo123** [comment]
Fine. But how do we deal with HDMI gone? Delete from config or wait for upstream changes ? File a bug report at armbian ?

**Comment by gpo123** [comment]
Actually I have a liitle problem with noise on BananaPi with 3,5mm jack on mainline kernel. This was not the case with legacy kernel. Did you observe the same ? I first thought it was a faulty cable or power supply but others seem to have this problem too as you can read here:
[https://forum.armbian.com/index.php/topic/397-mainline-a10a20-audio-driver/](url)
Do you know by chance how to convert *.dtb files into readable format ?

**Comment by gpo123** [comment]
I took the chance and updated cubietruck also to mainline 4.9. Running good so far:

![cubie_mainline_temp](https://cloud.githubusercontent.com/assets/9412171/22387343/6fa780f6-e4db-11e6-9527-7587d5eff97d.png)

These are the the processes running:
![cubie_mainline_htop](https://cloud.githubusercontent.com/assets/9412171/22387518/490a6f34-e4dc-11e6-963a-15b42320ffa3.png)

Quite good for this "deprecated" device ! Must prove itself now as stable and I am contended. Although HDMI is also missing.

**Comment by Fourdee** [comment]
> This was not the case with legacy kernel. Did you observe the same ?

Didn't check if i'am honest, I believe in the assumption of basic onboard audio support, then USB DAC/HAT's for serious audio usage. 
Although if it was fine on 3.x, its most likely an issue with the codec driver on 4.9. Seems the solution is per:
https://forum.armbian.com/index.php/topic/397-mainline-a10a20-audio-driver/#entry13933

> Do you know by chance how to convert *.dtb files into readable format ?

From compiled back to source? Nope sorry.

> Fine. But how do we deal with HDMI gone? Delete from config or wait for upstream changes ? File a bug report at armbian ?

https://forum.armbian.com/index.php/topic/397-mainline-a10a20-audio-driver/#entry24185

**Comment by Fourdee** [comment]
Tasks:
- 🈺 HDMI audio not available for 4.9 currently: https://forum.armbian.com/index.php/topic/397-mainline-a10a20-audio-driver/#entry24186
- 🈯️ Set BPi Pro to CARD 0 as default, 3.5mm support only
- 🈯️ Apply mixer settings when default card is selected: https://github.com/Fourdee/DietPi/issues/732#issuecomment-275702873
- 🈴 Set 4.9 image live (replacing 3.x). Need to wait for v143 release due to commit changes below.

I'll get these done.

**Comment by Fourdee** [comment]
Image live.
